### ISSUE TITLE
Using DbContext in Generated Code with using Roslyn

### ISSUE THREAD

**Issue by alpergunay** [dotnet/efcore#12647]
We have already opened an issue [https://github.com/dotnet/roslyn/issues/27381](url) to rosyln project but not sure the problem is related with the roslyn or entity framework. Below you can find the issue explanation and sample project. 

"We are using Roslyn and Microsoft.CodeAnalysis.CSharp libray to generate dynamic code and inject it one of our project. Simply, user is writing code and save it. Then roslyn is compiling the code, generate dll, create an instance and put it in memory. During tests we found that simple EF Linq queries generates wrong SQL when we use this approach. For example, we are passing a DbContext to created instance as a parameter and running a simple query, it generates wrong sql. Below, you can find the query, generated SQL and actual injected code"

Query :

```
var itemMasterId = (long)parameters[0];
        var query = from im in _dbContext.ItemMasters
                    where im.Id == 30
                    select new 
                    {
                        im.PalletType,
                        im.Description
                    };
        var result = query.FirstOrDefault();
```
Generated SQL :

```
SELECT [i].[ID], [i].[ABC_TYPE_ID], [i].[CATEGORY_ID], [i].[CHANGE_TIME], [i].[CHANGE_USER_ID], [i].[CODE], [i].[CREATE_TIME], [i].[CREATE_USER_ID], [i].[CUSTOMER_ACCOUNT_ID], [i].[DEFAULT_PACK_TYPE_ID], [i].[DESCRIPTION], [i].[GROUP_ID], [i].[IF_BUNDLE], [i].[IF_IMEI], [i].[IF_LOT_NO], [i].[IF_PART], [i].[IF_SERIAL_NO], [i].[IF_SHELF_LIFE], [i].[MINIMUM_ORDER_QUANTITY], [i].[NUMBER_OF_PARTS], [i].[PALLET_TYPE], [i].[RECORD_STATUS_ID], [i].[REF_CODE], [i].[ROW_VERSION], [i].[SAFETY_STOCK_QUANTITY], [i].[TYPE_ID], [i].[UNIT_OF_MEASURE_ID]
FROM [ITEM_MASTER] AS [i]
```
Code :

```
using SLOT.Infrastructure.DataLibrary.BusinessRuleEngine;
using SLOT.Infrastructure.DataLibrary.MasterData.ItemStowageTypeManagement;
using SLOT.Infrastructure.DataLibrary;
using SLOT.Infrastructure.DataModels.Common.Enums;
using Microsoft.EntityFrameworkCore;
using System.Linq;
using System;

[assembly: global::System.Runtime.Versioning.TargetFrameworkAttribute(".NETCoreApp,Version=v2.1", FrameworkDisplayName = "")

public class ZoneRule : BusinessRuleAbstract
{
    private SLOTDataContext _dbContext;

    public ZoneRule(SLOTDataContext dbContext)
    {
        _dbContext = dbContext;
    }

    public override object Execute(object[] parameters)
    {
        var itemMasterId = (long)parameters[0];
        var query = from im in _dbContext.ItemMasters
                    where im.Id == 30
                    select new 
                    {
                        im.PalletType,
                        im.Description
                    };
        var result = query.FirstOrDefault();
        
        return result.Description;
    }
}
```
In project you will find 3 test case with running the same query using entity framework core. Query includes only a Select with Group By. In first case, we are calling one method (GetItems) from directly in code which is in "ItemUnitOfWork" without using on the fly code and log the result. In second case, we are running the same query in on the fly code with generating library and creating an instance with using Roslyn APIs and log the results again. In third case, we are calling the method in "ItemUnitOfWork" again but in on the fly code with generating library and creating an instance.

Below you will find the test results. You will see the wrong generated SQL if we run query with using Roslyn APIs at second test case.

Test Results :

1. Code Is Called From StartUp

info: Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 2.1.0-rtm-30799 initialized 'RoslynTestDbContext' using provider Microsoft.EntityFrameworkCore.SqlServer' with options: None
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (99ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      ```
SELECT [x].[TYPE_ID] AS [Key], COUNT(*) AS [Count]
      FROM [ITEM_MASTER] AS [x]
      GROUP BY [x].[TYPE_ID]
```
Result:
1 -> 69
2 -> 13

2. Code Is Running on The Fly

info: Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 2.1.0-rtm-30799 initialized 'RoslynTestDbContext' using provider Microsoft.EntityFrameworkCore.SqlServer' with options: None
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (24ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      ```
SELECT [i].[ID], [i].[TYPE_ID]
      FROM [ITEM_MASTER] AS [i]
```
Result:
1 -> 69
2 -> 13

3.Code Is Called From On The Fly

info: Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 2.1.0-rtm-30799 initialized 'RoslynTestDbContext' using provider Microsoft.EntityFrameworkCore.SqlServer' with options: None
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (39ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      ```
SELECT [x].[TYPE_ID] AS [Key], COUNT(*) AS [Count]
      FROM [ITEM_MASTER] AS [x]
      GROUP BY [x].[TYPE_ID]
```
Result:
1 -> 69
2 -> 13

[RoslynTestApplication.zip](https://github.com/aspnet/EntityFrameworkCore/files/2187751/RoslynTestApplication.zip)

**Comment by divega** [comment]
EF Triage: Could you provide the logs generated during query executing at trace level?

**Comment by alpergunay** [comment]
I attached the screenshots of the trace log. 

![1](https://user-images.githubusercontent.com/25134503/42826018-16748f44-89ec-11e8-9c8f-3faac4fcf26c.jpg)
![2](https://user-images.githubusercontent.com/25134503/42826019-1694a59a-89ec-11e8-9c9d-7a7710062607.jpg)
![3](https://user-images.githubusercontent.com/25134503/42826021-16bc6530-89ec-11e8-88d6-f8db595b3879.jpg)

**Comment by alpergunay** [comment]
I attached also sample application. 
[RoslynTestApplication.zip](https://github.com/aspnet/EntityFrameworkCore/files/2202220/RoslynTestApplication.zip)

**Comment by smitpatel** [comment]
@alpergunay - The issue lies somewhere in the dynamically compiled code. For some reason instead of doing `IQueryable.GroupBy` it is using `IEnumerable.GroupBy` since Expression trees (which are required for SQL translation) works with IQueryables only, the query EF gets is raw `db.Items.ToList()` and everything after `GroupBy` never reaches EF. Hence you see difference in SQL.

And easy test to make this work without having EF SQL translation interfering is just printout `query.Expression` in all scenarios and they all should resolve to same output. (Then EF core will generate same query for each of them). In your sample app when I did this, I got following error.
```
System.ArgumentException: CS1061: 'IEnumerable<<anonymous type: long Key, int Count>>' does not contain a definition for 'Expression' and no extension method 'Expression' accepting a first argument of type 'IEnumerable<<anonymous type: long Key, int Count>>' could be found (are you missing a using directive or an assembly reference?)
```
Which clearly indicates that GroupBy caused transformation to Enumerable rather than preserving the Queryable.

I am not able to figure out why it is happening. Perhaps Roslyn team could help you figure out how to dynamically generate correct code in this case.

Closing this as external.
